File: ruby/muse/algorithms/number_theory/LeastCommonMultiple.rb
require "muse/algorithms/number_theory/GreatestCommonDivisor"

module LeastCommonMultiple
  def self.reduceToBinaryGCD(m, n)
    m = -m if m.negative?
    n = -n if n.negative?

    return 0 if m.zero? || n.zero?

    m / GreatestCommonDivisor.iterativeBinaryGCD(m, n) * n
  end

  def self.reduceToEuclidean(m, n)
    m = -m if m.negative?
    n = -n if n.negative?

    return 0 if m.zero? || n.zero?

    m / GreatestCommonDivisor.iterativeEuclidean(m, n) * n
  end
end

File: java/muse/algorithms/number_theory/LeastCommonMultiple.java
package muse.algorithms.number_theory;

public final class LeastCommonMultiple {

  private LeastCommonMultiple() {}

  public static long reduceToBinaryGCD(long m, long n) {
    if (m < 0) {
      m = -m;
    }
    if (n < 0) {
      n = -n;
    }

    if (m == 0 || n == 0) {
      return 0;
    }
    return m / GreatestCommonDivisor.iterativeBinaryGCD(m, n) * n;
  }

  public static long reduceToEuclidean(long m, long n) {
    if (m < 0) {
      m = -m;
    }
    if (n < 0) {
      n = -n;
    }

    if (m == 0 || n == 0) {
      return 0;
    }
    return m / GreatestCommonDivisor.iterativeEuclidean(m, n) * n;
  }
}
